google-cloud-firestore/src/main/java/com/google/cloud/firestore/TransactionOptions.java

```diff
@@ -83,10 +116,205 @@ public static TransactionOptions create(@Nonnull Executor executor) {
    * @param executor The executor to run the user callback code on.
    * @param numberOfAttempts The number of execution attempts.
    * @return The TransactionOptions object.
+   * @deprecated As of 2.0.0, replaced by {@link ReadWriteOptionsBuilder#setExecutor(Executor)} and
+   *     {@link ReadWriteOptionsBuilder#setNumberOfAttempts(int)}
+   * @see #readWriteOptionsBuilder()
    */
   @Nonnull
-  public static TransactionOptions create(@Nonnull Executor executor, int numberOfAttempts) {
-    Preconditions.checkArgument(numberOfAttempts > 0, "You must allow at least one attempt");
-    return new TransactionOptions(numberOfAttempts, executor);
+  @Deprecated
+  public static TransactionOptions create(@Nullable Executor executor, int numberOfAttempts) {
+    return readWriteOptionsBuilder()
+        .setExecutor(executor)
+        .setNumberOfAttempts(numberOfAttempts)
+        .build();
+  }
+
+  @Nonnull
+  public static ReadWriteOptionsBuilder readWriteOptionsBuilder() {
+    return Builder.readWriteBuilder();
+  }
+
+  @Nonnull
+  public static ReadOnlyOptionsBuilder readOnlyOptionsBuilder() {
+    return Builder.readOnlyBuilder();
+  }
+
+  public abstract static class Builder<T, B extends Builder<T, B>> {
+    @Nullable protected Executor executor;
+
+    protected Builder(@Nullable Executor executor) {
+      this.executor = executor;
+    }
+
+    @Nullable
+    public Executor getExecutor() {
+      return executor;
+    }
+
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    public B setExecutor(@Nullable Executor executor) {
+      this.executor = executor;
+      return (B) this;
+    }
+
+    @Nonnull
+    public abstract TransactionOptions build();
+
+    static ReadOnlyOptionsBuilder readOnlyBuilder() {
+      return new ReadOnlyOptionsBuilder(null, null);
+    }
+
+    static ReadWriteOptionsBuilder readWriteBuilder() {
+      return new ReadWriteOptionsBuilder(null, DEFAULT_NUM_ATTEMPTS);
+    }
+  }
+
+  public static final class ReadOnlyOptionsBuilder
+      extends Builder<ReadOnlyOptions, ReadOnlyOptionsBuilder> {
+    @Nullable private TimestampOrBuilder readTime;
+
+    private ReadOnlyOptionsBuilder(@Nullable Executor executor, @Nullable Timestamp readTime) {
+      super(executor);
+      this.readTime = readTime;
+    }
+
+    @Nullable
+    public TimestampOrBuilder getReadTime() {
+      return readTime;
+    }
+
+    public ReadOnlyOptionsBuilder setReadTime(@Nullable TimestampOrBuilder readTime) {
```

Comment:
Al of this needs documentation, which can probably just be copied from the Proto.